File: client/src/modules/assetmanager/schemas/holding/holding-cash-flow.schemas.ts
/**
 * Holding Cash Flow Schemas
 *
 * Zod validation schemas for HoldingCashFlow model.
 * Field names and validation rules match backend exactly (snake_case).
 *
 * Backend sources:
 * - Model: /server/apps/assetmanager/models/holding_models.py
 * - Schema: /server/apps/assetmanager/schemas/holding_schemas/holding_cash_flow_schemas.py
 * - Router: /server/apps/assetmanager/subrouters/holding_subrouters/holding_cash_flow_subrouter.py
 */

import { z } from 'zod';

// ==========================================
// Enums
// ==========================================

/**
 * Cash flow type options - matches backend CashFlowType enum
 * Backend: class CashFlowType(str, Enum)
 */
export const CashFlowTypeEnum = z.enum([
  'investment',
  'distribution',
  'dividend',
  'fee',
  'other',
]);

/**
 * Cash flow category options - matches backend CashFlowCategory enum
 * Backend: class CashFlowCategory(str, Enum)
 */
export const CashFlowCategoryEnum = z.enum([
  'actual',
  'projected',
]);

/**
 * Cash flow scenario options - matches backend CashFlowScenario enum
 * Backend: class CashFlowScenario(str, Enum)
 */
export const CashFlowScenarioEnum = z.enum([
  'actual',
  'budget',
  'forecast',
]);

// ==========================================
// HoldingCashFlow Schema (Full Representation)
// ==========================================

/**
 * HoldingCashFlow schema - full representation
 * Used for GET operations (single and list)
 *
 * Backend equivalent: class HoldingCashFlow(BaseModel)
 */
export const HoldingCashFlowSchema = z.object({
  id: z.number(),
  holding_id: z.number(),
  entity_id: z.number(),
  target_entity_id: z.number().nullable(),
  funding_round_id: z.number().nullable(),

  // Cash Flow Details
  date: z.string(),
  amount_debit: z.number(),
  amount_credit: z.number(),
  currency: z.string(),
  cash_flow_type: z.string(),
  category: z.string(),
  scenario: z.string(),

  // Transaction Reference
  cash_transaction_id: z.number().nullable(),

  // Additional Info
  transaction_reference: z.string().nullable(),
  description: z.string().nullable(),
  include_in_irr: z.boolean(),

  created_at: z.string(), // ISO datetime string from backend
  updated_at: z.string().nullable(), // ISO datetime string from backend
});

// ==========================================
// Input Schemas
// ==========================================

/**
 * Schema for creating a new holding cash flow (POST)
 * Excludes: id, created_at, updated_at (auto-generated)
 *
 * Backend equivalent: class HoldingCashFlowCreate(BaseModel)
 */
export const CreateHoldingCashFlowSchema = z.object({
  holding_id: z.number(),
  entity_id: z.number(),
  date: z.string(),
  cash_flow_type: CashFlowTypeEnum,
  amount_debit: z.number().optional(),
  amount_credit: z.number().optional(),
  currency: z.string().optional(),
  category: CashFlowCategoryEnum.default('actual'),
  scenario: CashFlowScenarioEnum.default('actual'),
  include_in_irr: z.boolean().optional(),
  target_entity_id: z.number().nullable().optional(),
  funding_round_id: z.number().nullable().optional(),
  cash_transaction_id: z.number().nullable().optional(),
  transaction_reference: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
});

/**
 * Schema for updating a holding cash flow (PUT/PATCH)
 * All fields optional to support partial updates
 *
 * Backend equivalent: class HoldingCashFlowUpdate(BaseModel)
 */
export const UpdateHoldingCashFlowSchema = z.object({
  holding_id: z.number().nullable().optional(),
  entity_id: z.number().nullable().optional(),
  target_entity_id: z.number().nullable().optional(),
  funding_round_id: z.number().nullable().optional(),

  // Cash Flow Details
  date: z.string().nullable().optional(),
  amount_debit: z.number().nullable().optional(),
  amount_credit: z.number().nullable().optional(),
  currency: z.string().nullable().optional(),
  cash_flow_type: CashFlowTypeEnum.nullable().optional(),
  category: CashFlowCategoryEnum.nullable().optional(),
  scenario: CashFlowScenarioEnum.nullable().optional(),

  // Transaction Reference
  cash_transaction_id: z.number().nullable().optional(),

  // Additional Info
  transaction_reference: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
  include_in_irr: z.boolean().nullable().optional(),
});

// ==========================================
// Type Exports
// ==========================================

export type CashFlowType = z.infer<typeof CashFlowTypeEnum>;
export type CashFlowCategory = z.infer<typeof CashFlowCategoryEnum>;
export type CashFlowScenario = z.infer<typeof CashFlowScenarioEnum>;
export type HoldingCashFlow = z.infer<typeof HoldingCashFlowSchema>;
export type CreateHoldingCashFlow = z.infer<typeof CreateHoldingCashFlowSchema>;
export type UpdateHoldingCashFlow = z.infer<typeof UpdateHoldingCashFlowSchema>;

// ==========================================
// Response Types
// ==========================================

/**
 * Response containing a single holding cash flow
 * Backend equivalent: class HoldingCashFlowResponse(BaseModel)
 */
export type HoldingCashFlowResponse = {
  success: boolean;
  data?: HoldingCashFlow;
  error?: string;
};

/**
 * Response containing multiple holding cash flows
 * Backend equivalent: class HoldingCashFlowsResponse(BaseModel)
 */
export type HoldingCashFlowsResponse = {
  success: boolean;
  data?: HoldingCashFlow[];
  error?: string;
};

File: server/apps/assetmanager/schemas/holding_schemas/holding_cash_flow_schemas.py
"""
HoldingCashFlow Schemas

Pydantic schemas for the HoldingCashFlow model following simplified schema guidelines.
"""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from datetime import date as date_type
from enum import Enum

# ==========================================
# Enum Types
# ==========================================

class CashFlowType(str, Enum):
    """Cash flow type options"""
    INVESTMENT = "investment"
    DISTRIBUTION = "distribution"
    DIVIDEND = "dividend"
    FEE = "fee"
    OTHER = "other"

class CashFlowCategory(str, Enum):
    """Cash flow category options"""
    ACTUAL = "actual"
    PROJECTED = "projected"

class CashFlowScenario(str, Enum):
    """Cash flow scenario options"""
    ACTUAL = "actual"
    BUDGET = "budget"
    FORECAST = "forecast"

# ==========================================
# HoldingCashFlow Schema (Full Representation)
# ==========================================

class HoldingCashFlow(BaseModel):
    """HoldingCashFlow schema - full representation"""
    id: int
    holding_id: int = Field(description="Associated holding ID")
    entity_id: int = Field(description="Associated entity ID")
    target_entity_id: int | None = Field(None, description="Target entity ID")
    funding_round_id: int | None = Field(None, description="Associated funding round ID")

    # Cash Flow Details
    date: date_type = Field(description="Cash flow date")
    amount_debit: float = Field(description="Debit amount")
    amount_credit: float = Field(description="Credit amount")
    currency: str = Field(description="Currency code")
    cash_flow_type: str = Field(description="Cash flow type")
    category: str = Field(description="Cash flow category")
    scenario: str = Field(description="Cash flow scenario")

    # Transaction Reference
    cash_transaction_id: int | None = Field(None, description="Associated security transaction ID")

    # Additional Info
    transaction_reference: str | None = Field(None, description="Transaction reference")
    description: str | None = Field(None, description="Description")
    include_in_irr: bool = Field(description="Include in IRR calculation")

    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)

# ==========================================
# Input Schemas
# ==========================================

class HoldingCashFlowCreate(BaseModel):
    """Schema for creating a new holding cash flow"""
    # Required fields
    holding_id: int = Field(description="Associated holding ID")
    entity_id: int = Field(description="Associated entity ID")
    date: date_type = Field(description="Cash flow date")
    cash_flow_type: CashFlowType = Field(description="Cash flow type")

    # Fields with defaults
    amount_debit: float = Field(default=0, description="Debit amount")
    amount_credit: float = Field(default=0, description="Credit amount")
    currency: str = Field(default="USD", max_length=3, description="Currency code")
    category: CashFlowCategory = Field(default=CashFlowCategory.ACTUAL, description="Cash flow category")
    scenario: CashFlowScenario = Field(default=CashFlowScenario.ACTUAL, description="Cash flow scenario")
    include_in_irr: bool = Field(default=True, description="Include in IRR calculation")

    # Optional fields
    target_entity_id: int | None = Field(None, description="Target entity ID")
    funding_round_id: int | None = Field(None, description="Associated funding round ID")
    cash_transaction_id: int | None = Field(None, description="Associated security transaction ID")
    transaction_reference: str | None = Field(None, description="Transaction reference")
    description: str | None = Field(None, description="Description")

class HoldingCashFlowUpdate(BaseModel):
    """Schema for updating a holding cash flow"""
    holding_id: int | None = None
    entity_id: int | None = None
    target_entity_id: int | None = None
    funding_round_id: int | None = None

    # Cash Flow Details
    date: date_type | None = None
    amount_debit: float | None = None
    amount_credit: float | None = None
    currency: str | None = Field(None, max_length=3)
    cash_flow_type: CashFlowType | None = None
    category: CashFlowCategory | None = None
    scenario: CashFlowScenario | None = None

    # Transaction Reference
    cash_transaction_id: int | None = None

    # Additional Info
    transaction_reference: str | None = None
    description: str | None = None
    include_in_irr: bool | None = None

# ==========================================
# Response Types
# ==========================================

class HoldingCashFlowResponse(BaseModel):
    """Response containing a single holding cash flow"""
    success: bool
    data: HoldingCashFlow | None = None
    error: str | None = None

class HoldingCashFlowsResponse(BaseModel):
    """Response containing multiple holding cash flows"""
    success: bool
    data: list[HoldingCashFlow] | None = None
    error: str | None = None
